### PR TITLE
fix(botocore): keep newlines in json data for kinesis records

### LINKED ISSUE #4317: `\n` stripped from kinesis JSON records when distributed tracing is enabled
### Which version of dd-trace-py are you using?

1.4.1 through `datadog-lambda-python` (https://github.com/DataDog/datadog-lambda-python/blob/main/poetry.lock#L111)

### Which version of pip are you using?

What's used to build `arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Python38:62` layer.

### Which version of the libraries are you using?

What's referenced on `arn:aws:lambda:us-east-1:464622532012:layer:Datadog-Python38:62` layer.

See [datadog-lambda-python](https://github.com/DataDog/datadog-lambda-python).

### How can we reproduce your problem?

1. Create a lambda that can write to a kinesis stream using `boto3`
2. Use the [serverless](https://github.com/DataDog/serverless-plugin-datadog) plugin with default configuration.
3. Deploy the lambda
4. Run the lambda and write the following on the target kinesis stream: 

```python
kinesis_client = boto3.client('kinesis')
data = { 'foo': 1 }
kinesis_client.put_records(
  StreamName='my_stream', 
  Records=[{ 'Data': bytearray(f'{json.dumps(data)}\n', 'utf-8') }]
)
```

### What is the result that you get?

The record actually written to the kinesis stream does not include the `\n` at the end. 

### What is the result that you expected?

The record actually written to the kinesis stream should include the `\n` at the end. 

### Why is this an issue?

Our kinesis streams are ingested through Firehose and written to S3. From S3, they are read from databricks. Databricks (and other like athena), expect JSON records to be separated by a new line (see: https://spark.apache.org/docs/latest/sql-data-sources-json.html and https://jsonlines.org/). That's why we are currently writing them with a `\n` at the end. 

After adding the DataDog lambda layer, our pipeline broke because the `\n` was no longer honored, as it's discarded when read here: https://github.com/DataDog/dd-trace-py/blob/b4449028209854654291cb762d02baf4bd1cf8cd/ddtrace/contrib/botocore/patch.py#L168 

The `\n` to separate JSONs is a common enough use case that has been recently supported by AWS natively when using [Firehose dynamic partitioning](https://docs.aws.amazon.com/firehose/latest/dev/dynamic-partitioning.html#dynamic-partitioning-new-line-delimiter), which regrettably we can't use at this time.

We could add a separate transformation step from firehose => s3 and append the newline, but the behavior was unexpected, as we didn't assume the tracer would be changing our kinesis streams data payloads, and regrettably the issue went undetected for a while.

Opening the ticket to check if this use case is something you would want to consider and properly support, and discuss alternatives (from adding a configuration option to append the new line, to check for it during decoding and attempt to preserve, or providing hooks to decode/encode the message + tracing data).

### PR DESCRIPTION
**Pull request by hghotra**
## Description

During our serialization & de-serialization of kinesis records, any `\n` at the end of the record is stripped by the json library. Customer pipelines rely on this `\n` delimiter to distinguish between different records. This change appends the line break to the end of the record string if it was originally there.

## Checklist
- [ ] Followed the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) when writing a release note.
- [ ] Add additional sections for `feat` and `fix` pull requests.
- [ ] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Motivation

Add support for Databricks which relies on the `\n` delimiter to be present at the end of the record. Look at the related issue for more details.

## Design 

If the incoming record has a `\n` at the end of the string, we add the `\n` into the string post serialization since it was stripped during the de-serialization process by the json library.

## Testing strategy

- two unit tests have been added

```
pytest -k test_kinesis_put_records_newline_base64_trace_injection tests/contrib/botocore/test.py
pytest -k test_kinesis_put_records_newline_json_trace_injection tests/contrib/botocore/test.py
```

## Relevant issue(s)

Fixes #4317

## Reviewer Checklist
- [x] Title is accurate.
- [x] Description motivates each change.
- [x] No unnecessary changes were introduced in this PR.
- [x] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [x] Release note has been added and follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines), or else `changelog/no-changelog` label added.
- [x] All relevant GitHub issues are correctly linked.
- [x] Backports are identified and tagged with Mergifyio.

### CODE REVIEW

**Comment by brettlangdon** [comment]
Would you be able to add a release note for this?

https://github.com/DataDog/dd-trace-py/blob/1.x/docs/contributing.rst#generating-a-release-note

**Review by IvanTopolcic — APPROVED** [review]
LGTM!

File: releasenotes/notes/fix-kinesis-newline-strip-issue-3a2419b6f29fcab8.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#    * Use full sentences with sentence-casing and punctuation.
+#    * Before using Datadog specific acronyms/terminology, a release note must first introduce them with a definition.
+#
+#    Release notes should not:
+#    * Be vague. Example: ``fixes an issue in tracing``.
+#    * Use overly technical language
+#    * Use dynamic links (``stable/latest/1.x`` URLs). Instead, use static links (specific version, commit hash) whenever possible so that they don't break in the future.
+fixes:
+  - |
+    This fix resolves an issue where any `\n` at the end of a kinesis record is stripped by the json library.
```
Comment (brettlangdon) [inline releasenotes/notes/fix-kinesis-newline-strip-issue-3a2419b6f29fcab8.yaml:29]:
```suggestion
    botocore: This fix resolves an issue where any ``\n`` at the end of a kinesis record is stripped by the json library.
```

**Comment by brettlangdon** [comment]
@Mergifyio backport 1.6

File: releasenotes/notes/fix-kinesis-newline-strip-issue-3a2419b6f29fcab8.yaml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#    * Be concise
+#    * Include actionable steps with the necessary code changes
+#    * Include relevant links (bug issues, upstream issues or release notes, documentation pages)
+#    * Use full sentences with sentence-casing and punctuation.
+#    * Before using Datadog specific acronyms/terminology, a release note must first introduce them with a definition.
+#
+#    Release notes should not:
+#    * Be vague. Example: ``fixes an issue in tracing``.
+#    * Use overly technical language
+#    * Use dynamic links (``stable/latest/1.x`` URLs). Instead, use static links (specific version, commit hash) whenever possible so that they don't break in the future.
```
Comment (majorgreys) [inline releasenotes/notes/fix-kinesis-newline-strip-issue-3a2419b6f29fcab8.yaml:26]:
```suggestion
```

File: ddtrace/contrib/botocore/patch.py [hunk 3]
```diff
@@ -165,8 +168,17 @@ def inject_trace_to_eventbridge_detail(params, span):
         entry["Detail"] = detail_json
 
 
+def get_json_from_str(data_str):
+    # type: (str) -> Tuple[str, Optional[Dict[str, Any]]]
+    data_obj = json.loads(data_str)
+
+    if data_str.endswith(LINE_BREAK):
```
[the remark this answers is not in the document]
Comment (hghotra) [reply ddtrace/contrib/botocore/patch.py:175]: No, as long as we append one to the end of the record, we should be okay. If the customer comes back and this is a concern, we can take another look.

File: releasenotes/notes/fix-kinesis-newline-strip-issue-3a2419b6f29fcab8.yaml (hunk 1 above)
Comment (majorgreys) [inline releasenotes/notes/fix-kinesis-newline-strip-issue-3a2419b6f29fcab8.yaml]:
```suggestion
    botocore: Before this change, the botocore integration stripped newlines from the JSON string encoded in the data blob of Amazon Kinesis records. This change includes a terminating newline if it is present in the decoded data.
```

File: releasenotes/notes/fix-kinesis-newline-strip-issue-3a2419b6f29fcab8.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+---
+fixes:
+  - |
+    botocore: This fix resolves an issue where any ``\n`` at the end of a kinesis record is stripped by the json library.
```
Comment (brettlangdon) [inline releasenotes/notes/fix-kinesis-newline-strip-issue-3a2419b6f29fcab8.yaml:4]:
```suggestion
    botocore: Before this change, the botocore integration stripped newlines from the JSON string encoded in the data blob of Amazon Kinesis records. This change includes a terminating newline if it is present in the decoded data.
```

**Review by brettlangdon — APPROVED** [review]

File: ddtrace/contrib/botocore/patch.py [hunk 5]
```diff
@@ -165,8 +168,17 @@ def inject_trace_to_eventbridge_detail(params, span):
...
+        return LINE_BREAK, data_obj
+    return None, data_obj
```
Comment (mabdinur) [inline ddtrace/contrib/botocore/patch.py:177]: typing nit. We skip type checking on modules in `ddtrace/contrib` so mypy missed this. 

```suggestion
    return "", data_obj
```

**Review by Yun-Kim — DISMISSED** [review]

**Review by majorgreys — DISMISSED** [review]

File: ddtrace/contrib/botocore/patch.py [hunk 6]
```diff
@@ -216,11 +229,16 @@ def inject_trace_to_kinesis_stream_data(record, span):
         return
 
     data = record["Data"]
-    data_obj = get_kinesis_data_object(data)
+    line_break, data_obj = get_kinesis_data_object(data)
     data_obj["_datadog"] = {}
     HTTPPropagator.inject(span.context, data_obj["_datadog"])
+
```
Comment (mabdinur) [inline ddtrace/contrib/botocore/patch.py:235]:
```suggestion
nit: we can remove this line
```

**Review by mabdinur — DISMISSED** [review]
Left a small nit but looks good to me

**Review by Yun-Kim — DISMISSED** [review]

**Review by Yun-Kim — APPROVED** [review]

**Review by mabdinur — APPROVED** [review]

**Pull request merged**